Add medical-pre-authorization sample agent

agarwaldipali-hash opened the pull request:
This PR introduces the sample code for the Pre-Auth AI Agent, built using the Google ADK framework.

This agent showcases an automated solution for the healthcare pre-authorization process. It intelligently analyzes medical records and insurance policies to determine eligibility for medical treatments.

The live demo is available at: https://demo-1382.1987984870407.demo.altostrat.com/

This sample code provides a powerful, real-world example of the Google ADK framework in action. It serves as a practical blueprint for developers and customers, especially in the healthcare industry, demonstrating how to build sophisticated AI agents to solve complex business problems.

jackwotherspoon reviewed (commented):
@agarwaldipali-hash I wonder if there is a better name than `pre-auth-ai-agent`?

I wonder if something like `medical-insurance` or `medical-pre-authorization` is a better name for the agent. As someone just looking at the list of agents I have no clue what `pre-auth-ai-agent` entails from the name...

jackwotherspoon commented:
> @jackwotherspoon I have renamed it to medical-pre-authorization-ai-agent.

Is there a reason you are wanting to keep `ai-agent` appended to the name?

I recommended `medical-pre-authorization` due to the fact that these samples are in the `agents` folder, so it is implied they are all agents... and `ai` is not needed as agents imply AI by their nature.

I think `medical-pre-authorization` is better than `medical-pre-authorization-ai-agent`

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/prompt.py:
```diff
@@ -0,0 +1,45 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def return_root_agent_instruction() -> str:
```
What is the purpose of this being a function?

Why can't we just have it be a string and then import it directly? 

Although it should probably be all capitals to signify a constant (i.e. `AGENT_INSTRUCTION`)

https://github.com/google/adk-samples/blob/8faee5f027ff454120da2f54e4449f4ffd077227/python/agents/software-bug-assistant/software_bug_assistant/prompt.py#L15

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
```diff
@@ -0,0 +1,250 @@
+1.  **Prerequisites**
+
+    *   Python 3.11+
+    *   Poetry
```
We are trying to slowly move the samples over to use `uv` instead of Poetry, if it is not too much trouble can you update to use uv? It looks like you already have a `uv.lock` file... which makes life easier.

See https://github.com/google/adk-samples/tree/main/python/agents/data-science#prerequisites

agarwaldipali-hash replied:
Made the changes to use uv

agarwaldipali-hash replied:
updated uv.lock

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
```diff
@@ -0,0 +1,250 @@
+
+    *   Set up Google Cloud credentials.
+
+        *   You may set the following environment variables in your shell, or in
+            a `.env` file instead.
```
should we link to the example .env file here?

agarwaldipali-hash replied:
done

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
````diff
@@ -0,0 +1,250 @@
+Hello, who are you?
+```
+
+Sampled responses of these requrests are shown below in the [Example
````
```suggestion
Sampled responses of these requests are shown below in the [Example
```

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
````diff
@@ -0,0 +1,250 @@
+For running tests and evaluation, install the extra dependencies:
+
+```bash
+poetry install --with dev
````
if we decide to go with `uv` commands these can easily be updated to use `uv run`

agarwaldipali-hash replied:
Done

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/requirements.txt:
This file is not used anywhere is it?

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/subagents/data_analyst/prompt.py:
```diff
@@ -0,0 +1,75 @@
+the patient's OptiCare Health Plans policy and does not meet the strict medical necessity criteria for coverage.
+"""
+
+def return_instructions_data_analyst():
```
again, not sure why we need a function here...? We should just use a constant...

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/subagents/information_extractor/agent.py:
```diff
@@ -0,0 +1,30 @@
+information_extractor= Agent(
+   model='gemini-2.5-flash',
+   name="information_extractor",
+   description="Agent that extracts the details on user insurance policy and medical necessity for a pre-authorization request from documents provided by the user.",
```
nit: max line length of 80, use `"""` to spread this over multiple lines for readability

agarwaldipali-hash replied:
made the required changes

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/subagents/information_extractor/prompt.py:
```diff
@@ -0,0 +1,27 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def return_instructions_information_extractor():
```
nit: use a constant, no need for a function here...

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/subagents/information_extractor/tools/tools.py:
```diff
@@ -0,0 +1,222 @@
+    model = "gemini-2.5-flash"
+
+    # Construct a prompt for the Gemini model to extract the treatment name
+    si_text = f"""
```
what does `si_text` mean? Is this some type of jargon...?

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/subagents/information_extractor/tools/tools.py:
```diff
@@ -0,0 +1,222 @@
+    
+
+    # Define the text part of the prompt
+    msg1_text1 = types.Part.from_text(text=f"""
```
why is this called `msg1_text1`?

jackwotherspoon commented on python/agents/medical-pre-authorization/medical_pre_authorization/subagents/information_extractor/tools/tools.py:
```diff
@@ -0,0 +1,222 @@
+
+
+    # Define the text part of the prompt
+    msg1_text1 = types.Part.from_text(text="""You are an AI assistant specialized in analyzing medical reports. 
```
ditto here

jackwotherspoon reviewed (changes requested):
Did another pass.

a) Majority of files need a newline at EOF. You can configure this automatically in your editor of choice usually.
b) Please add an entry for you new agent to the top level README https://github.com/google/adk-samples/tree/main/python/agents within this PR
c) Update READMDE instructions to use `uv`, it seems you have a `uv.lock` file yet README mentions Poetry....

agarwaldipali-hash commented:
@jackwotherspoon Made below three changes as per our conversation on Gchat:
1. Updated all the commands to use uv run
2. Removed "pip install uv". It links to official documentation for uv installation
3. Updated to use uv run --extra deployment deployment/deploy.py --create
4. Verified that all the instructions work as expected
5. Removed all poetry commands

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
```diff
@@ -0,0 +1,249 @@
+    # Verify if uv is installed correctly
+    uv --version
+
+
+    # Create a virtual environment and sync dependencies.
+    uv venv
```
`uv venv` is not needed, the venv is automatically created by `uv sync`

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
````diff
@@ -0,0 +1,249 @@
+
+## Running Tests
+
+For running tests and evaluation, install the extra dependencies:
+
+```bash
+uv pip install -e .[dev]
+```
````
This is not needed if we use `uv` in an efficient way

agarwaldipali-hash replied:
Done

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
````diff
@@ -0,0 +1,249 @@
+the `pytest` module:
+
+```bash
+uv run python3 -m pytest tests
+uv run python3 -m pytest eval
````
Uv makes it super easy to run tests with the `dev` extra deps...

```suggestion
uv run --extra dev pytest tests
uv run --extra dev pytest eval
```

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
````diff
@@ -0,0 +1,249 @@
+If you forgot the AGENT_ENGINE_ID, you can list existing agents using:
+
+```bash
+uv run python3 deployment/deploy.py --list
````
Proper `uv` command:

```suggestion
uv run --extra deployment deployment/deploy.py --list
```

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
````diff
@@ -0,0 +1,249 @@
+
+You may interact with the deployed agent using the `test_deployment.py` script
+```bash
+$ export USER_ID=<any string>
+$ export AGENT_ENGINE_ID=<AGENT_ENGINE_ID>
+$ uv run python3 deployment/test_deployment.py --resource_id=${AGENT_ENGINE_ID} --user_id=${USER_ID}
````
Leading `$` will make these commands error...
```suggestion
export USER_ID=<any string>
export AGENT_ENGINE_ID=<AGENT_ENGINE_ID>
uv run --extra deployment deployment/test_deployment.py --resource_id=${AGENT_ENGINE_ID} --user_id=${USER_ID}
```

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
```diff
@@ -0,0 +1,249 @@
+$ export USER_ID=<any string>
+$ export AGENT_ENGINE_ID=<AGENT_ENGINE_ID>
+$ uv run python3 deployment/test_deployment.py --resource_id=${AGENT_ENGINE_ID} --user_id=${USER_ID}
+Found agent with resource ID: ...
+Created session for user ID: ...
+Type 'quit' to exit.
+Input: Hello, what can you do for me?
+Response: Hello! I'm a pre-authorization agent. I can help you process pre-authorization requests for medical treatments. I'll need some details about the treatment, your medical records, and your health insurance policy to do so.
```
this should be moved to a separate code block from the input...

You shouldn't mix input that the user needs to copy with output...

agarwaldipali-hash replied:
Done

jackwotherspoon commented on python/agents/medical-pre-authorization/README.md:
````diff
@@ -0,0 +1,249 @@
+To delete the deployed agent, you may run the following command:
+
+```bash
+uv run python3 deployment/deploy.py --delete --resource_id=${AGENT_ENGINE_ID}
````
```suggestion
uv run --extra deployment deployment/deploy.py --delete --resource_id=${AGENT_ENGINE_ID}
```

jackwotherspoon reviewed (changes requested):
Final pass, once addressed we can get this merged in 😄
